sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+properties, but this design document does not attempt to address them all.
+Rather, it describes the design for each "category" of use and illustrates each
+category with a specific feature.  For example `accessor` is used to illustrate
+properties that are applied to a kernel argument, but the same technique could
+be used for other variables that are captured as kernel arguments.
+
+In all cases, the goal of this design is to explain how a DPC++ program that
+uses properties is consumed by the device compiler and eventually represented
+in LLVM IR.  This typically involves some logic in the header files which
+results in a C++ annotation that contains the properties.  The device compiler
+front-end is responsible for consuming this annotation and producing some
+corresponding LLVM IR.  One of the goals of this design is to avoid changes to
+the front-end each time we add a new property, so the front-end is not required
+to understand each property it consumes.  Instead, it follows a mechanical
+process for converting properties listed in the C++ annotation into LLVM IR,
+and this mechanical process need not be updated when we add new properties.
+
+Once the information about properties is represented in IR, it is available to
+compiler passes.  For example, the `sycl-post-link` tool might use a property
+in order to perform one of its transformations.  Some properties are consumed
+by the DPC++ compiler, but others are transformed into SPIR-V for use by the
+JIT compiler.  This design document also describes how this SPIR-V
+transformation is done.
+
+
+## Properties on a global variable type
+
+One use for compile-time properties is with types that are used exclusively
+for declaring global variables.  One such example is the
+[SYCL\_EXT\_ONEAPI\_DEVICE\_GLOBAL][2] extension:
+
+[2]: <extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc>
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The following code illustrates a `device_global` variable that is declared with
+two compile-time properties:
+
+```
+using sycl::ext::oneapi;
+
+device_global<int,
+  property_list_t<
+    device_image_scope::value_t,
+    host_access::value_t<host_access::access::read>>>
+  dm1;
+```
+
+The header file represents these properties with an internal C++ attribute
+named `[[__sycl_detail__::add_ir_global_variable_attributes()]]` whose value
+is a list that is created through a template parameter pack expansion:
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template <typename T, typename ...Props>
+class
+#ifdef __SYCL_DEVICE_ONLY__
+  [[__sycl_detail__::add_ir_global_variable_attributes(
+    Props::meta_name..., Props::meta_value...
+    )]]
```

Comment:
What is the expected behavior for runtime property types here? Will all properties (runtime and compile-time) have to have `meta_name` and `meta_value`?

Alternatively, would it make sense to use a trait instead? Say we have a trait `PropertyMetaInfo` as
```c++
template<typename T> struct PropertyMetaInfo {
  static constexpr const char *meta_name = "";
  static constexpr const char *meta_value = "";
};
template<typename PropertyT, typename... PropertyValueTs>
struct PropertyMetaInfo<property_value<PropertyT, PropertyValueTs...>> {
  static constexpr const char *meta_name = PropertyMetaInfo<PropertyT>::meta_name;
  static constexpr auto meta_value = PropertyMetaInfo<PropertyT>::meta_value;
};
```
we can use
```c++
  [[__sycl_detail__::add_ir_global_variable_attributes(
    PropertyMetaInfo<Props>::meta_name..., PropertyMetaInfo<Props>::meta_value...
    )]]
```
Then have `__sycl_detail__::add_ir_global_variable_attributes` ignore all entries where the `meta_name` is an empty string. This way any properties that don't translate to attributes won't have to bother with setting meta information.

Reply:
> What is the expected behavior for runtime property types here? Will all properties (runtime and compile-time) have to have meta_name and meta_value?

I think we decided that all properties to `property_list` are compile-time properties.  A property may have a runtime (or compile-time) parameter, but the property itself is always compile-time.

> Alternatively, would it make sense to use a trait instead?

Yes, I think it's OK to have a separate trait instead if that's easier.

> Then have `__sycl_detail__::add_ir_global_variable_attributes` ignore all entries where the meta_name is an empty string.

This seems problematic, though.  Properties that have no parameter are currently defined (in this design doc) to use the empty string as a placeholder for the parameter value.

> This way any properties that don't translate to attributes won't have to bother with setting meta information.

Is there some metaprogramming way to exclude certain properties from the pack expansion that populates the C++ attribute's parameter list?  I agree that that there will be cases where a class that uses properties will want only some of those properties to be included in the attribute.

Reply:
> This seems problematic, though. Properties that have no parameter are currently defined (in this design doc) to use the empty string as a placeholder for the parameter value.

That only applies to `meta_value`, right? `meta_name` would have to be some relevant attribute.

> Is there some metaprogramming way to exclude certain properties from the pack expansion that populates the C++ attribute's parameter list?

Potentially, but I don't know of any simple way of doing it.

Reply:
> That only applies to meta_value, right? meta_name would have to be some relevant attribute.

Yes, that's correct.  I have another concern, though, about setting `meta_name` to the empty string as a way to indicate that the property should not have an IR attribute.  This technique assumes that all classes that use the property want the same behavior.  However, it seems possible that two classes might might both use property `P`.  One class might want `P` to correspond to an IR attribute while the other class does not.  Therefore, I think it would better if we could find some metaprogramming construct that let each class filter the set of properties that contribute to the C++ attribute's parameter list.

>> Is there some metaprogramming way to exclude certain properties from the pack expansion that populates the C++ attribute's parameter list?
>
> Potentially, but I don't know of any simple way of doing it.

@rolandschulz is my go-to person for metaprogramming questions like this!

Reply:
> > That only applies to meta_value, right? meta_name would have to be some relevant attribute.
> 
> Yes, that's correct. I have another concern, though, about setting `meta_name` to the empty string as a way to indicate that the property should not have an IR attribute. This technique assumes that all classes that use the property want the same behavior. However, it seems possible that two classes might might both use property `P`. One class might want `P` to correspond to an IR attribute while the other class does not. Therefore, I think it would better if we could find some metaprogramming construct that let each class filter the set of properties that contribute to the C++ attribute's parameter list.
>

We could probably just have a trait for each class or groups of classes, then specialize only for the ones that translate to a relevant attribute. It is a little more work for whoever adds a new property, but for these cases you have to specify the separation anyway.

Alternatively we can have one trait that also has the receiving class as a template parameter, a little like `is_property_of`. It doesn't allow for simple groupings, say you want an attribute for multiple classes that always use the same attributes. I don't imagine that to be a huge problem though.

Reply:
In both of these cases, the ramification on the C++ attribute is the following, right?  It's possible that some of the parameters corresponding to property names are the empty string (`""`).  When this happens, the front-end should skip this property (and its value) and not generate an IR attribute.  Correct?

If this is easier than some metaprogramming solution that filters out properties from the parameter list, then I am OK with this solution.

FWIW, I don't think this design document needs to document the exact way the header files work.  The main thing I wanted to illustrate with the headers is that we can use a parameter pack expansion based on the properties to populate the arguments of a C++ attribute.  I don't think this document needs to spell out exactly how that pack is formed, whether from metadata inside each property or via some other trait.  I'll have to think about how to change the wording in this document to be a bit less proscriptive.

Reply:
> In both of these cases, the ramification on the C++ attribute is the following, right? It's possible that some of the parameters corresponding to property names are the empty string (""). When this happens, the front-end should skip this property (and its value) and not generate an IR attribute. Correct?

That's my thinking, yeah. It makes the AST a little uglier and the behavior of the attribute a little more branched, but I think for our sake this is the easiest way around it. That said, I am curious as to whether @rolandschulz has ideas on how to do the filtering.

Reply:
I was thinking more about this, and I have a new proposal for how we can filter the properties for all of these C++ attributes.  The idea is to shift some of the work to the CFE, but I think this will significantly reduce the effort in the headers.  Tagging @premanandrao and @elizabethandrews for their comments about the CFE impact.

To recap, the problem is that the application may pass many properties to the property list of some class.  However, we may not want all of these properties to be translated into IR attributes, which is what will happen if the header looks like this:

```
template <typename T, typename ...Props>
class
#ifdef __SYCL_DEVICE_ONLY__
  [[__sycl_detail__::add_ir_global_variable_attributes(
    Props::meta_name..., Props::meta_value...   // This pack expansion will include all properties that are specified for this class
    )]]
#endif
  device_global<T, property_list<Props...>> {/*...*/};
```

According to @rolandschulz, there is a way to filter the properties using meta-programming, but it seems involved and requires creating a second class.  As an alternative, I think we could add another parameter to the `[[__sycl_detail__::add_ir_global_variable_attributes()]]` attribute which specifies a "filter list" of property names.  For example:

```
template <typename T, typename ...Props>
class
#ifdef __SYCL_DEVICE_ONLY__
  [[__sycl_detail__::add_ir_global_variable_attributes(
    {"sycl-device-image-scope", "sycl-host-access"},  // This is a filter applied to the properties
    Props::meta_name..., Props::meta_value...
    )]]
#endif
  device_global<T, property_list<Props...>> {/*...*/};
```

When the CFE consumes this C++ attribute, it constructs a "pass list" from the strings in the first parameter.  When reading the remaining parameters, the CFE compares each property name with the "pass list".  If the property name is NOT on the pass list, it ignores that property (and its value), and does NOT generate any IR attribute for that property.

Since there will be many cases where we do not need a "pass list", I'd also like to make this first parameter optional.  Therefore, if the first parameter of the C++ attribute is not a braced list of strings, the CFE doesn't do any filtering of the properties.

This seems simpler than the "empty string" approach that @steffenlarsen proposes, and I think the resulting code in the headers will be more self-documenting.

Comments?

Reply:
That would be neat, but I don't know if the infrastructure is there for attributes to accept list arguments like this, so I worry that it may complicate the implementation of the new attributes. That said, I do agree this would look a lot cleaner. Luckily I may be wrong and CFE folks can correct me. 😄

Reply:
Yes, I was also wondering if it would be difficult to add support for an attribute parameter that is a brace-enclosed list.  If that is hard, we could instead express the "pass list" as a count following by a variable number of strings like:

```
template <typename T, typename ...Props>
class
#ifdef __SYCL_DEVICE_ONLY__
  [[__sycl_detail__::add_ir_global_variable_attributes(
    2,                                              // This is the number of elements in the "pass list" filter
    "sycl-device-image-scope", "sycl-host-access",  // These strings are the "pass list"
    Props::meta_name..., Props::meta_value...
    )]]
#endif
  device_global<T, property_list<Props...>> {/*...*/};
```

That's not as nice looking, though, so I'd prefer to use the brace list if that is possible.

Reply:
I don't see any documented clang attributes accepting brace-enclosed arguments, but from looking at the code that parses this, it seems it should be accepted.  Could we tentatively go with this option, and come back to option 2 if the CFE runs into issues during implementation?

Reply:
Yes, we can do that.  We can always tweak the design doc if we find problems during implementation.

Reply:
I am okay with that solution, though as a passing note; trying to use brace-enclosed arguments in an attribute that accepts expressions (`clang::annotate` in this case) results in a parser error `expected expression`. (https://godbolt.org/z/YnG1MhfPP)

Reply:
I documented this proposal in 44763693264bc91488fc5c8772b61f0cc5d9210b.  See especially the new section "Filtering properties".

If we decide later that we cannot easily add the braced list, we can update the design doc.

Reply:
I'm happy to take a look, but I likely won't get to it until Monday at the earliest.

Reply:
I was able to review this a bit and have some concerns.

As @steffenlarsen points out, this is all pretty novel for attributes. Some attributes already support expressions, but no attribute currently expects an init-list argument and so who knows how well this will work in practice. I could not find a situation where an init list is accepted currently, so there's definitely no reason to think this will work out of the box without some effort on our part.

FWIW, this design also makes the attribute unergonomic in some ways. e.g.,
```
[[attr(foo, {1, 2, 3})]] void whatever(); // expected-error {{silly, you forgot to include <initializer_list> and C++ is a language only a mother could love}}
```

Reply:
We continued this discussion in another thread in this same PR.  See https://github.com/intel/llvm/pull/4937#discussion_r768083995.

The summary is that we will see if it is difficult to implement the braced list syntax.  If it turns out to be too difficult, we have a fall-back proposal.